samples/99_zil_interpreter/app/parser.rb

```diff
@@ -0,0 +1,375 @@
+      buf_char = @buffer[@buf_index]
+      @buf_index += 1
+
+      if buf_char == '\\' # observed in this--> <STRING !\" !,WBREAKS>
+        next_char = peek
+        if next_char == '"'
+          atom += read_char
+        end
+      elsif ATOM_CHARS.include?(buf_char)
+        atom += buf_char
+      else
+        @buf_index -= 1
+        break
+      end
+    end
+
+    atom.to_sym
+  end
+
+  def read_char
+    char = @buffer[@buf_index]
+    @buf_index += 1
+    char
+  end
+
+  def read_comment
+    # read comment char, then read the expresson that follows
+    semi = read_char # read in ;
+    comment = read_expression # read commented out stuff and ignore
+
+    Syntax::Comment.new(comment)
+  end
+
+  # %<...>
+  def read_cond_macro
+    pct_char = read_char # read '%'
+    expr = read_expression # read '<...>'
+
+    Syntax::Macro.new(expr)
+  end
+
+  def read_decl
+    hashtag = read_char # read #
+    decl = read_atom # read DECL
+    skip_whitespace
+    list = read_list
+
+    Syntax::Decl.new(*list)
+  end
+
+  def read_expression
+    debug_flag = false
+    @expr_depth += 1
+    raise "Expression recursion overflow" if @expr_depth > 50
+
+    expr_char = peek
+    if expr_char == '<' #  Starts with '<' then read_form
+      log "+ FORM " if debug_flag
+      expr = read_form
+    elsif expr_char == '(' #  Starts with '(' then read_list
+      log "+ LIST " if debug_flag
+      expr = read_list
+    elsif expr_char == '"' #  Starts with '"' then read_string
+      log "+ STRING " if debug_flag
+      expr = read_string
+    elsif expr_char == ';' #  Starts with ';' then read_comment
+      log "+ COMMENT " if debug_flag
+      expr = read_comment
+    elsif expr_char == '#' && peek_string(5) == '#DECL'
+      log "+ DECL " if debug_flag
+      expr = read_decl
+    elsif expr_char == '%' && peek_string(6) == '%<COND'
+      log "+ COND MACRO " if debug_flag
+      expr = read_cond_macro
+    elsif expr_char == "'"
+      log "+ MACRO " if debug_flag
+      expr = read_macro
+    elsif is_fix?(expr_char) # Starts with 0123456789 then read_fix
+      log "+ FIX " if debug_flag
+      expr = read_fix
+    else
+      #  else read atom
+      log "+ ATOM " if debug_flag
+      expr = read_atom
+    end
+
+    @expr_depth -= 1
+    expr
+  end
+
+  def read_fix
+    # UC1: 123456
+    # UC2: *3777*
+    # UC3: #2 0111
+
+    fix = ""
+
+    # step 1: just load the string based on the three types
+    while @buf_index < @buffer.length - 1 # iterate characters in line
+      buf_char = @buffer[@buf_index]
+      @buf_index += 1
+
+      if FIX_CHARS.include?(buf_char)
+        #log "[TMP] " + buf_char
+        fix += buf_char
+      elsif buf_char == " " && fix == '#2'
+        fix += buf_char
+      else
+        @buf_index -= 1
+        break
+      end
+    end
+
+    # step 2: convert to number
+    if fix.start_with?('*')
+      fix = fix.gsub('*', '')
+      fix = fix.to_i(base=8)

```

Comment:
a few redundant assignments imho, (also why `base=8`? you never use that)

Reply:
Base 8 is octal - it's used in gverbs.zil in Zork 1, 2, and 3. 

gverbs.zil:
  `<PRINTN <BAND <GET 0 1> *3777*>>`

Reply:
ie. the `base=8` is redundant, `String#to_i` takes one optional argument just `8` will suffice

Reply:
I reworked the redundant stuff.  base8 is required.

Reply:
For readability, `base=8` is super clear what's going on.